icicle/include/api/gen_c_api.py

```diff
@@ -0,0 +1,89 @@
+#!/usr/bin/env python3
+"""
+This script generates extern declarators for every curve/field
+"""
+
+from pathlib import Path
+
+API_PATH = Path(__file__).resolve().parent
+TEMPLATES_PATH = API_PATH.joinpath("templates")
+
+CURVES_MODULES = {
+    "curve",
+    "msm",
+    "ecntt"
+}
+
+FIELDS_MODULES = {
+    "field",
+    "ntt",
+    "poseidon",
+}
+
+CURVES_CONFIG = {
+    "bn254": {},
+    "bls12_381": {},
+    "bls12_377": {},
+    "bw6_761": {},
+    "grumpkin": {"ecntt", "g2", "ntt"},
+}
+
+FIELDS_CONFIG = {
+    "babybear": {"poseidon"}
+}
+
+WARN_TEXT = """\
+// WARNING: This file is auto-generated by a script.
+// Any changes made to this file may be overwritten.
+// Please modify the code generation script instead.
+// Path to the code generation script: icicle/include/api/gen_c_api.py
+
+
+"""
+
+if __name__ == "__main__":
+
+    # Generate API for ingo_curve
+    for curve, skip in CURVES_CONFIG.items():
+        for module in CURVES_MODULES | FIELDS_MODULES - set(skip):
+            module_templates_path = TEMPLATES_PATH.joinpath(module)
+
+            module_path = API_PATH.joinpath(curve).joinpath(module)
+            module_path.mkdir(parents=True, exist_ok=True)
+
+            for header_template in module_templates_path.iterdir():
+                if "ext" in header_template.name and "ext" in skip:
+                    continue
+
+                if "g2" in header_template.name and "g2" in skip:
+                    continue
+
+                contents = open(header_template).read()
```

Comment:
should use 'with open...' instead of open() directly, especially since you don't close the files.

Reply:
why?

Reply:
`with open` auto closes the file once the with block ends.

Reply:
yeah, I know, but why do we need it here?:)

Reply:
It's generally recommended to use context-managers when handling resoruce such as files in Python. It's more robust since it closes the file even when an exception is raised. It's also clearer since you can see the scope where the file is used and you don't have to close it explicitly.
I think that calling open() directly should only be done when you cannot use a context manager.